text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+
+```rust
+// an empty async function
+async fn drop_async<T>(to_drop: T) { }
+
+fn poll_drop_ready<T>(&mut self, cx: &mut Context<'_>) -> Poll<()>
+    where T: ?Sized
+{
+    // implemented through a lang item
+}
+```
+
+The `drop_async` function is analogous to the `drop` function, and is also
+added to the prelude as well. This function drops the value in an async
+context, guaranteeing that its `poll_drop_ready` method will be called.
+
+The `poll_drop_ready` function calls this value's "drop ready glue" - it calls
+`poll_drop_ready` on this value and also on all of its fields, recursively. The
+exact ordering and semantics of this glue are specified in the reference
+section.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Guarantees about when `poll_drop_ready` is called
+
+In general, users cannot assume that `poll_drop_ready` will ever be called,
+just as they cannot assume that destructors will run at all. However, they can
+assume less about `poll_drop_ready` than they can about `drop`, and that should
+be noted.
+
+In particular, users cannot safely assume that their values will be dropped in
+an async context, regardless of how they are intended to be used. It's this
+easy to supress the `poll_drop_ready` call of a value:
+
+```rust
+async fn ignore_poll_drop_ready<T>(x: T) {
+    // by passing `x` to `drop`, a non-async context, `poll_drop_ready` is
+    // never called when dropping `x`.
+    drop(x);
+}
+```
+
+However, we do guarantee that values dropped in an async context *will* have
+`poll_drop_ready` called, and we even guarantee the drop order between
+variables and between fields of a type. When the user sees a value go out of
+scope *in an async context*, they know that `poll_drop_ready` is called. And we
+do guarantee that the destructors of all fields of that type will be called.
+
+## Changes to destructor calls in async contexts
+
+In async contexts, we change the generated destructor calls to insert, before
+each call, this code:
+
+```rust
+while let Poll::Pending = mem::poll_drop_ready(obj, cx) {
+    yield Poll::Pending;
+}
+```
+
+This "prepares" the object to be destroyed by calling its destructor, calling
+`poll_drop_ready` on the object and all of its fields. This is called before
+any destructors are run on that object.
+
+In non-async contexts, no change is made to destructors. This means that users
+writing low level primitives which may want to run async destructor code on
+values they are destroying (such as inside the poll method of a
+manually-implemented future) will need to call `mem::poll_drop_ready`
+themselves, or this code will be skipped.
+
+Therefore, there are two cases in which users will have to take special care to
+ensure that async destructors are called:
+
+1. If they would already have to call `drop_in_place` to ensure that normal
+   destructors are called (as in data structures like `Vec`).
```

Comment:
would there be an `async_drop_in_place`, or does the user need to manually perform the poll?

```rust
unsafe {
    futures_util::future::poll_fn(|cx| poll_drop_ready(&mut *ptr, cx)).await;
    drop_in_place(ptr);
}
```

Reply:
I'm not sure what the best approach here is, but possibly this would be helpful. Then we would probably want to change the drop glue description from being as described to instead being described as calling async_drop_in_place in an async context. This would also save a virtual call in the case of trait objects.

Note that for types like vec, for example, the right approach will be to call `poll_drop_ready` on the slice in Vec's poll_drop_ready method. Slices will have to have drop glue implemented to join a poll_drop_ready on all of their members.